src/Block/Service/AbstractBlockService.php

```diff
@@ -31,33 +32,62 @@ abstract class AbstractBlockService implements BlockServiceInterface
     protected $name;
 
     /**
+     * NEXT_MAJOR: Remove this property.
+     *
      * @var EngineInterface|null
      */
     protected $templating;
 
     /**
-     * @param EngineInterface|string $templatingOrDeprecatedName
+     * @var Environment
+     */
+    private $twig;
+
+    /**
+     * NEXT_MAJOR: Make `$twig` argument mandatory and remove other arguments.
+     *
+     * @param Environment|EngineInterface|string $templatingOrDeprecatedName
      */
-    public function __construct($templatingOrDeprecatedName = null, EngineInterface $templating = null)
+    public function __construct($templatingOrDeprecatedName = null, EngineInterface $templating = null, Environment $twig = null)
```

Comment:
Why do we need the third parameter?

Reply:
Because we want to respect the BC promise for users which are extending this class and need to keep the behavior for `getTemplating()` method. At least, that was my reasoning for adding this argument instead of just using the argument 1.

Reply:
I prefer using the first parameter only, because we will drop the third parameter with the next major release. So this is just a short living solution.

IMHO we should use the first parameter for this. If the user passes an `Environment` as first parameter, the `getTemplating` method should throw a `RuntimeException`.

Reply:
It's fine for me. Could we use `BadMethodCallException` instead of `RuntimeException`?

Reply:
Your comments were addressed.